lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -631,6 +631,42 @@ void swift::checkActorIsolation(const Expr *expr, const DeclContext *dc) {
             return { false, expr };
           }
         }
+
+        // Actor-isolated properties can't be passed inout to async functions.
+        // If the function being called is async, we run checkMemberReference
+        // on all of the inout parameters to ensure that they do not belong to
+        // an actor.
+        FunctionType *fnType = apply->getFn()->getType()->getAs<FunctionType>();
+        if (fnType && fnType->isAsync()) {
+          // Since we get different types for the arg list depending on
+          // whether the function is called with one parameter or multiple
+          // parameters, just collect the inout expressions so that we don't
+          // have extra special-casing
+          llvm::SmallVector<Expr *, 3> inoutArgs;
+          if (ParenExpr *arg = dyn_cast<ParenExpr>(apply->getArg())) {
+            if (InOutExpr *inoutArg = dyn_cast<InOutExpr>(arg->getSubExpr()))
+              inoutArgs.push_back(inoutArg->getSubExpr());
+          } else if (TupleExpr *argList =
+                         dyn_cast<TupleExpr>(apply->getArg())) {
+            inoutArgs.reserve(argList->getNumElements());
+            for (Expr *arg : argList->getElements()) {
+              if (InOutExpr *inoutArg = dyn_cast<InOutExpr>(arg))
+                inoutArgs.push_back(inoutArg->getSubExpr());
+            }
```

Comment:
We might have to look through `ParenExpr` within each argument (e.g., for `f((&self.state))`). I recommend moving the collection of `inout` parameters out into a helper function.

Reply:
`f((&self.state))` already emits an error, though it's about an extraneous `&`, which isn't exactly clear. I could just strip the parens off too, but if it's invalid does it matter?